pilot/pkg/security/authn/v1alpha1/policy_applier.go

```diff
@@ -103,25 +116,84 @@ func outputLocationForJwtIssuer(issuer string) string {
 	return locationPrefix + fmt.Sprintf("%x", sum)
 }
 
-// ConvertPolicyToJwtConfig converts policy into Jwt filter config for envoy.
-func convertPolicyToJwtConfig(policy *authn_v1alpha1.Policy) *jwtfilter.JwtAuthentication {
-	policyJwts := collectJwtSpecs(policy)
-	if len(policyJwts) == 0 {
-		return nil
+func convertToEnvoyJwtConfig(policyJwts []*authn_v1alpha1.Jwt) *envoy_jwt.JwtAuthentication {
+	var requirements []*envoy_jwt.JwtRequirement
+	providers := map[string]*envoy_jwt.JwtProvider{}
+	for _, policyJwt := range policyJwts {
+		provider := &envoy_jwt.JwtProvider{
+			Issuer:               policyJwt.Issuer,
+			Audiences:            policyJwt.Audiences,
+			ForwardPayloadHeader: outputLocationForJwtIssuer(policyJwt.Issuer),
+			Forward:              true,
+			PayloadInMetadata:    policyJwt.Issuer,
+		}
+
+		for _, location := range policyJwt.JwtHeaders {
+			provider.FromHeaders = append(provider.FromHeaders, &envoy_jwt.JwtHeader{
+				Name: location,
```

Comment:
To support RcToken, we will also need to treat `ingress-authorization` header slightly different. I.e, you may need to set `value_prefix` if the header is `ingress-authorization`

Reply:
It seems the prefix is not fixed and istio proxy just trims whatever before the whitespace: https://github.com/istio/proxy/blob/master/src/envoy/http/jwt_auth/token_extractor.cc#L89-L92

@quanjielin Is this by design? I feel it's better to have an explicit prefix.

The other question is who is going to set the `ingress-authorization` header in the JWT authentication API? Do we require user to specify it by manually or do we just silently add this to the JWT filter config?

Reply:
As for the other question, I think the users need to set `ingress-authorization` explicitly, right? Envoy Jwt filter cannot differentiate regular jwt vs RCtoken, unless someone specifies the header and prefix, whereas Istio jwt filter can differentiate based on the prefix `' '`.

Reply:
My guess is user only need to specify high-level config in the API, the header and prefix is not in the API and is actually implementation detail. @diemtvu @quanjielin

Reply:
Yes the users only need to specify it in their policy config. We can add a header field in Istio API. Envoy already has header and prefix fields. Since we're adding new field to the Istio API, we need to make sure to convert to Istio jwt config (rather than Envoy jwt config) when users use the old API (with the one with no header field).

Reply:
Ideally, we may need to change our Istio Jwt API to provide prefix option. After that, all user need to set it explicitly to use rctoken. For now, maybe it's ok for pilot to do that implicitly if `ingress-authorization` is used. User still need to specify JWT location from `ingress-authorization` header.

Regarding Istio proxy implementation, I agree it's not strong enough. As we will strengthen it with this migration, we can leave it for now, imo.

Reply:
+1 to add prefix into jwt policy api (to align with envoy config), could be done later.

Reply:
@diemtvu What's the prefix should be used for `ingress-authorization`? I thought this is not decided yet as Istio proxy trims what ever before the whitespace.

One hack way is to set the prefix to a single whitespace `" "` which works based on the implementation detail of the Envoy JWT filter, but I don't think this is a right approach for us and it doesn't for every cases.

Reply:
+  @quanjielin Do you know what the value it is?
On the second thought, I think we may want to add prefix to our JWT API to make it fully customizable.

Reply:
a sample: 
```
ingress-authorization: Istio eyJhbGciOiJFUzI1Ni***
```

Reply:
Marking resolved for now, will revisit to add support once the API is finalized.

Reply:
We need to handle it here, before we have new (Istio) API, otherwise existing RcToken will break.
A simplest one is if the location is `ingress-authorization` header, then also set prefix to 'Istio'

Reply:
I see, updated to support RCToken.